scipy/optimize/tests/test_least_squares.py

```diff
@@ -787,6 +788,72 @@ def test_basic():
     assert_allclose(res.x, 0, atol=1e-10)
 
 
+def test_callback():
+    # test that callback function works as expected
+
+    results = []
+    
+    def my_callback_optimresult(intermediate_result: OptimizeResult):
+        results.append(intermediate_result)
+        
+    def my_callback_x(x):
+        r = OptimizeResult()
+        r.nit = 1
+        r.x = x
+        results.append(r)
+        return False
+        
+    def my_callback_optimresult_stop_exception(intermediate_result: OptimizeResult):
+        results.append(intermediate_result)
+        raise StopIteration
+        
+    def my_callback_optimresult_stop_true(intermediate_result: OptimizeResult):
+        results.append(intermediate_result)
+        return True
+        
+    def my_callback_x_stop_exception(x):
+        r = OptimizeResult()
+        r.nit = 1
+        r.x = x
+        results.append(r)
+        raise StopIteration
+        
+    def my_callback_x_stop_true(x):
+        r = OptimizeResult()
+        r.nit = 1
+        r.x = x
+        results.append(r)
+        return True
+    
+    # Try for different function signatures and stop methods
+    callbacks_nostop = [my_callback_optimresult, my_callback_x]
+    callbacks_stop = [my_callback_optimresult_stop_exception,
+                      my_callback_optimresult_stop_true,
+                      my_callback_x_stop_exception,
+                      my_callback_x_stop_true]
+    
+    # Try for all the implemented methods: trf, trf_bounds and dogbox
+    calls = [lambda callback: least_squares(fun_trivial, 5.0, method='trf', callback=callback),
+             lambda callback: least_squares(fun_trivial, 5.0, method='trf', bounds=(-8.0, 8.0), callback=callback),
+             lambda callback: least_squares(fun_trivial, 5.0, method='dogbox', callback=callback)]
+    
+    for my_callback in callbacks_nostop:
+        for call in calls:
+            results.clear()
+            res = call(my_callback)  # Call the different implemented methods
+            assert_(len(results) > 0)   # Check that callback was called
+            assert_(results[-1].nit > 0)  # Check that results data makes sense
+            assert_(res.status != -2)  # Check that it didn´t stop because of the callback
+    
+    for my_callback in callbacks_stop:
+        for call in calls:
+            results.clear()
+            res = call(my_callback)  # Call the different implemented methods
+            assert_(len(results) > 0)   # Check that callback was called
+            assert_(results[-1].nit == 1)  # Check that only one iteration was run
+            assert_(res.status == -2)  # Check that it stopped because of the callback
```

Comment:
minor point -- we now prefer plain `assert` over the older `assert_` for *new* testing functions